core/integration/directory_test.go

```diff
@@ -926,3 +927,63 @@ func TestDirectorySync(t *testing.T) {
 		require.Equal(t, []string{"foo"}, entries)
 	})
 }
+
+func TestDirectoryEntriesIncludeAndRecursive(t *testing.T) {
+	t.Parallel()
+
+	c, ctx := connect(t)
+
+	srcDir := c.Directory().
+		WithNewFile("main.go", "").
+		WithNewFile("func.go", "").
+		WithNewFile("test.md", "").
+		WithNewFile("foo.txt", "").
+		WithNewFile("README.txt", "").
+		WithNewDirectory("/subdir").
+		WithNewFile("/subdir/foo.txt", "").
+		WithNewFile("/subdir/README.md", "").
+		WithNewDirectory("/subdir2").
+		WithNewDirectory("/subdir2/subsubdir").
+		WithNewFile("/subdir2/baz.txt", "").
+		WithNewFile("/subdir2/TESTING.md", "").
+		WithNewFile("/subdir/subsubdir/package.json", "").
+		WithNewFile("/subdir/subsubdir/index.mts", "").
+		WithNewFile("/subdir/subsubdir/JS.md", "")
+
+	t.Run("include only markdown", func(t *testing.T) {
+		entries, err := srcDir.Entries(ctx, dagger.DirectoryEntriesOpts{Include: "*.md"})
+
+		require.NoError(t, err)
+		require.ElementsMatch(t, entries, []string{"test.md"})
+	})
+
+	t.Run("recursive listing", func(t *testing.T) {
+		entries, err := srcDir.Entries(ctx, dagger.DirectoryEntriesOpts{Recursive: true})
+
+		require.NoError(t, err)
+		require.ElementsMatch(t, entries, []string{
+			"func.go", "main.go", "test.md", "foo.txt", "README.txt",
+			"subdir", "subdir/foo.txt", "subdir/README.md",
+			"subdir2", "subdir2/subsubdir", "subdir2/baz.txt", "subdir2/TESTING.md",
+			"subdir/subsubdir", "subdir/subsubdir/package.json",
+			"subdir/subsubdir/index.mts", "subdir/subsubdir/JS.md",
+		})
+	})
+
+	t.Run("recursive that include only markdown", func(t *testing.T) {
+		entries, err := srcDir.Entries(ctx, dagger.DirectoryEntriesOpts{Recursive: true, Include: "*.md"})
+
+		require.NoError(t, err)
+		require.ElementsMatch(t, entries, []string{
+			"test.md", "subdir/README.md",
+			"subdir2/TESTING.md", "subdir/subsubdir/JS.md",
+		})
+	})
```

Comment:
I'm not actually convinced this belongs in `Entries` (happy to be told I'm wrong though).

Today `Entries` returns the contents of a target directory - with these new options, it instead does something different - it tells you all the files that match a given glob. It feels like two different functionalities squashed into a single call depending on the parameters.

I wonder if we could instead add a new method, `Glob` (or similar)? This would just return a list of all the files that match the given glob - as a nice plus, we can also directly use `IncludePatterns` from upstream buildkit (though we still need to call `ReadDir` recursively, unless we want to upstream something like that into BuildKit).

Reply:
> Or we could always assume Recursive when Include is set - which would be functionally equivalent to the above, it's just a naming convention as to whether it's under Glob or under Entries.

`Recursive` and `Include` are 2 different option, if I want `*.md`, I do not expected a recursivity. However I can add a check to see if the glob contains `**` and then activate `Recursive` in that way, this could works.

> IncludePatterns from upstream buildkit (though we still need to call ReadDir recursively, unless we want to upstream something like that into BuildKit).

See https://github.com/dagger/dagger/pull/5824#discussion_r1337617427

> Today Entries returns the contents of a target directory - with these new options, it instead does something different - it tells you all the files that match a given glob. It feels like two different functionalities squashed into a single call depending on the parameters.

I'm not sure it's right to put it in another method, this would mostly lead to code duplication, if an user want to get the content of the target directory, he can also expect to get subdirectories as mentioned in #5685.

WDYT @vito @helderco @gerhard

Reply:
I agree with @jedevc; I don't think we need `Recursive`. `*.md` should always match the top level, and `**/*.md` should be how you specify recursive globs.

My suggestion API wise is this:

```graphql
extend type Directory {
  entries(include:[String!], exclude: [String!]): [String!]!
}
```

This would match the various include/exclude filter APIs we have everywhere else.

I can see the appeal of having a separate `Glob` API instead, but having separate include/exclude filters implies both are optional, so you might as well just put it on `Entries` IMO. (Unless there's some sort of built-in 'negated glob' syntax I'm not aware of.)

Reply:
`IncludePattern` Only expect a string, I'm not sure we can support multiple include / exclude.
In our case excluse will have to be done manually too, I would prefer supporting that in a further PR

I added your suggestion about `Recursive`deletion in [4364df0](https://github.com/dagger/dagger/pull/5824/commits/4364df05a7849233a953d8c94155e72a985a08cc)

Reply:
> In our case excluse will have to be done manually too, I would prefer supporting that in a further PR

Agreed about future PR for now. However - we should instead upstream the support for this to buildkit (should be a relatively simple fix that we can merge quickly, happy to pair on it :tada:), instead of adding custom logic.